xdsl/backend/csl/lowering/convert_stencil_to_csl.py

```diff
@@ -0,0 +1,399 @@
+    layout_module: CslModuleOp
+    program_sym_name: StringAttr
+
+    program_params: dict[str, ParamOp] = {}
+    layout_params: dict[str, ParamOp] = {}
+    program_imports: dict[str, ImportModuleConstOp] = {}
+    layout_imports: dict[str, ImportModuleConstOp] = {}
+    program_vars: dict[str, Operation] = {}
+    layout_vars: dict[str, Operation] = {}
+
+    def add_param_to_program(self, param: ParamOp):
+        self.program_module.body.block.add_op(param)
+        self.program_params[param.param_name.data] = param
+
+    def add_param_to_layout(self, param: ParamOp):
+        self.layout_module.body.block.add_op(param)
+        self.layout_params[param.param_name.data] = param
+
+    def add_import_module_to_program(self, lib: ImportModuleConstOp):
+        self.program_module.body.block.add_op(lib)
+        self.program_imports[lib.module.data] = lib
+
+    def add_import_module_to_layout(self, lib: ImportModuleConstOp):
+        self.layout_module.body.block.add_op(lib)
+        self.layout_imports[lib.module.data] = lib
+
+    def add_params_to_program(self, *params: ParamOp):
+        for param in params:
+            self.add_param_to_program(param)
+
+    def add_params_to_layout(self, *params: ParamOp):
+        for param in params:
+            self.add_param_to_layout(param)
+
+    def add_import_modules_to_program(self, *lib: ImportModuleConstOp):
+        for l in lib:
+            self.add_import_module_to_program(l)
+
+    def add_import_modules_to_layout(self, *lib: ImportModuleConstOp):
+        for l in lib:
+            self.add_import_module_to_layout(l)
+
+    def add_var_to_program(self, name: str, op: Operation):
+        self.program_module.body.block.add_op(op)
+        self.program_vars[name] = op
+
+    def add_var_to_layout(self, name: str, op: Operation):
+        self.layout_module.body.block.add_op(op)
+        self.layout_vars[name] = op
+
+
+def generate_program_module(ctx: TranslationContext) -> None:
+    """
+    Generates the basic structure of a program module needed for stencil computations.
+    This function should *not* translate the actual stencil computation, but set up everything
+    needed for these, including imports, params, and top-level constants.
+    Params are generated to match `@set_tile_code` invocation of the
+    layout module. Also imports memcpy, utils, and stencil_comms modules. Vars/consts
+    are set up as required to perform these imports.
+    """
+
+    # program params to be supplied by layout module
+    ctx.add_params_to_program(
+        s_params := ParamOp.get(Named.stencil_comms_params, ComptimeStructType()),
+        m_params := ParamOp.get(Named.memcpy_params, ComptimeStructType()),
+        ParamOp.get(Named.is_border_region_pe, IntegerType(1)),
+    )
+
+    # size of the z dimension
+    ctx.add_var_to_program(
+        Named.z_dim,
+        z_dim := Constant(
+            IntegerAttr(ctx.grid_dim[2], IntegerType(16, Signedness.UNSIGNED))
```

Comment:
Is there a reason why this is explicitly marked unsigned? See https://mlir.llvm.org/docs/Rationale/Rationale/#integer-signedness-semantics for more info on signedness.

Reply:
(see discussion above)